packages/react/src/createHooks.ts

```diff
@@ -0,0 +1,116 @@
+import { Query, Context, UpdateType, INode, LiveResult } from '@aphro/runtime-ts';
+import counter from '@strut/counter';
+import { useCallback, useMemo, useRef, useSyncExternalStore } from 'react';
+import { suspend } from 'suspend-react';
+
+const count = counter('model-infra/CreateHooks');
+
+export type ContextPromise = Promise<Context>;
+
+export type UseQueryOptions = {
+  // If you want to limit what sorts of updates will cause the live query to re-fire
+  on?: UpdateType;
+  // If you want to cache the results of the query.
+  // This'll allow components to unmount and remount without losing previously queried state.
+  // Should we just compute the key for them?
+  // We can given the declarative nature of the queries...
+  key: string;
+};
+
+type QueryReturnType<Q> = Q extends Query<infer M> ? M : any;
+
+export function createHooks(contextPromise: ContextPromise) {
+  // suspends the subtree using a hook that depends on context initialization
+  // until the context has been initialized.
+  function suspendOnContext() {
+    return suspend(() => {
+      return contextPromise;
+    }, ['context']);
+  }
+
+  function useQueryContext() {
+    const ctx = suspendOnContext();
+    return ctx;
+  }
+
+  function useQuery<Q extends Query<QueryReturnType<Q>>>(
+    queryProvider: (ctx: Context) => Q,
+    deps: any[],
+    { on, key }: UseQueryOptions,
+  ) {
+    const ctx = useQueryContext();
+
+    // allow non-stable query provider functions without rerunning dependent effects -
+    // dependencies are manually managed by the deps array parameter.
+    const providerRef = useRef(queryProvider);
+    providerRef.current = queryProvider;
+    const stableProvider = useCallback((ctx: Context) => providerRef.current(ctx), deps);
```

Comment:
Why do we need to copy the provider into a ref then create a callback? It seems like this would work with just the `queryProvider` and not wrapping it in a `stableProvider` -- given we don't do any comparisons on the stableProvider later.

Reply:
Hm, this may be some leftover fluff from my experiments trying to make this stable and avoid the loop I described in the PR. I could try pulling it out.

Reply:
It looks like a no-op now and leftover from a prior iteration.